### ISSUE TITLE
Create IA players

### ISSUE THREAD

**Issue by koburk** [steemnova/steemnova#207]
This might be a bit hard but challenging and Imo, very interesting. We could have tons of IA players to fight with (More places to attack to start and they can attack too, so be aware!).

Steps:

Create IA Player
IA Players will have preferences such as:
- Bunker makers (Big % of their income will be spent on defense), not much time spent ONLINE
- Agressive play (Big % of their income will be spent on fleet), a lot of time spent ONLINE.
- Hybrid play (Income will be spent on defences and fleet).
- NoobStyle (random stuff O.o)
 noobstyle might change into another style after some time (Everyone learns to play!).
- Enable IA Players to build structures
- Enable IA Players to research
- Enable IA Players to build fleet
- Enable IA Players to build defenses
- Enable IA Players to colonize.
Based on IA style, the colonies will be placed in different positions to fit its play style.
- Enable IA Players to spy.
- Spy report will be processed and IA Randomness will decide if attack or not.
- Enable IA Players to attack
- Enable IA Players to recicle.
Based on IA Player experience, reciclers will be sent before/during/after the attack.
- Enable IA Players to do FleetSaving.
- Enable IA Players to comerce (ok, this one might be the hardest hehe)
- Enable IA Players to request comerce to human players.
- Enable IA Players to do attacks with other IA Players.
- Enable IA Players to destroy moons.
Aggresive style will try to hunt players down. Tons of variables will ve involved, such as IA Player points, profits after destroing the enemy fleet, etc.
- Enable IA Players to use phalanx.
- Enable IA Players to attack fleets using phalanx information.
- Generate an IA Player ranks based on power and experience.

Cheers

**Comment by koburk** [comment]
Working on machine learning ...

**Comment by gregariomansa** [comment]
To be able to create an use bots, be it AI or not, would be wonderful for the Steemnova community. Some years ago, little before de facto quitting the game, I wrote about this. I think it was a suggestion to @mys. He kindly replied, but with a negative answer.

**Comment by gregariomansa** [comment]
Since this is open source, this could be implemented outside Steemnova. Now I remember that is exactly what a suggested to @mys. Something like "If you don't do it, maybe I will".

So, @kuburk, if you start a server that allows bots, you can count me in.

**Comment by koburk** [comment]
Thank you for your reply. Actually I'm already working on that!. Actually, instead of using the steemnova engine for testing (because this engline was made to work using php as "brain"), i'm developing a full new engine, using php only to display info and making all the "thinking" within SQL Side.
Once I have it runnin, I'll have to see how (if possible) to translate my development into steemnova engine.

Identifying IA as such is gonna be easy, but the goal is to set their IA to "play" not to be farms. That's the main reason i'm developing the new engine to allow machine learning in the process (Yes, I might be going too deep into this, but I like the challenge). In time we're gonna be learning from bots for sure.

![image](https://user-images.githubusercontent.com/60184425/73654060-ff3f6f80-468a-11ea-93e0-f5f8806926bd.png)

There's an example of my environment values. Ofcourse there's still more things to add to allow bigger events, those are my bassics for now.

If you pay attention, you'll see "max planets", and "min/max fields", instead of using ranges for planet position like ogame, i will use randomness to set each planet's fields. The difficulty in this case will be get the biggest possible, because the first positions will have a very low chance to be that big, same for the furthest. I still need to check the math for that.... So many things to check haha

**Comment by mys** [comment]
In my opinion the best place to set AI is in PHP Cronjob. Cron script can iterate over bot-players specified in database and take actions for them. All in random interval.
The biggest challenge is to precise path of actions bot should do.

**Comment by koburk** [comment]
You're right, but having something in php calling sql or just having sql, I much rather the second option. Barely no performance issues but because you can have all the code in the same place. At the end that "cron" can be scheduled in mysql too.
I work as DBA, so that's why I might be a bit commited to use a bit more the mysql power instead of using php to iterate or take actions, those are way faster using pure sql (Focused on big amounts of information).

**Comment by Qwatayean** [comment]
It depends of how many bots should be executed. One per cron or multiple
per cron?

If is one per cron well it's playable, but one player per cron you barely
notice his progress in-game (mainly if is not cron executed per minute).

A class to be executed per player request could be a good choice. Basically
player would be playing as "2" players (im at buildings, one random bot
randomly builds, im at fleet, one random bot randomly fleets). It's better
to have a minimal task exectued on player request rather than an
exhaustive cronjob. A bot cronjob can eat up to 10-30 secs on run, and no
one wants to wait such time. If is 5 minutes per cron, then it's ok, 20
seconds in 5 minutes its barely noticeable, but per minute, not the best
approach.

Source: I have a cronjob bot mode, and working on a request one which has
shown better results.

koburk <notifications@github.com> escreveu no dia segunda, 3/02/2020 à(s)
12:31:

**Comment by koburk** [comment]
I'm not 100% sure, it was a bit hard for me to really get what you were trying to say, correct me and my apologies in advance if i'm not getting your idea...

Seems like you are trying to figure out the best way to improve performance, but there's no much to be worried about. During testing, i was able to get 10.000 IA players runing at the same time, checking for actions ever 1 sec with over 97% of success
What does that 97% of success means?, basically it means you can run 9700 orders every 1 sec without worries about if those were executed, but that's a full push try. Normally we wouldn't be using as muchas CPU/Memory trying to simulate every 1 sec, instead every 1 minute to hours between them (the difference would be the IA play style). Basically if we get to play vs 10.000 IA players, not everysingle one of them will be doing stuff at the same time, but having a % of performing an action.

The global/general idea is: you ask those 10.000 IA players: "are you online?" if it's online you ask again: "are you going to do something?", those two questions are set to enable actions or not, then if those two answers were YES, you ask: "do you wanna try something new or follow machine learning stuff?", based on that: "what do you want to do?"

that's a very brief idea.

**Comment by Qwatayean** [comment]
No problem! What I was trying to say is the game have some demanding
queries to process in-game development such assign building to queue,
trying to send ships, update planet resource, etc...

If you found a way to counter all of this and have a responsive bot/AI
system, man, that's perfect.

koburk <notifications@github.com> escreveu no dia segunda, 3/02/2020 à(s)
14:43:

**Comment by koburk** [comment]
Indeed, the queries are not the same as user queries (Not stored, processed instead).
I'm glad I could answer without changing topic hehe.

**Comment by koburk** [comment]
At this point, my IA_Players are able to build stuff... Here you can see my log (From newer to older descending order).
Basically every minute, the server checks the questions I've posted before (is online, wants to do something ...), then checks what it can actually do (based on requirements, levels and resources), and then builds. Not much until now, but step by step.


DB Object ia_connection manages ia_brain object
DB Object ia_connection (Event) created
DB Object ia_brain manages queue_lock object
DB Object object_level created
DB Object wants_to_do (Function) created
DB Object ia_brain (Stored procedure) created
DB Object ia_queue created
DB Object is_online (Function) created
DB Object object_detail created
DB Object object_name created
DB Object universe created
DB Object rapid_fire created
DB Object requriments created
DB Object planets created
DB Object create_user manages full_galaxy object.
DB Object create_user (Stored procedure) created
DB Object change_log created
DB Object users created
DB Object ia_style_description created
DB Object types created
Random names for IA_Players
Change log created

**Comment by koburk** [comment]
Update:
My IA-Dumb-brain-dead-Players are already fighting each other to climb my ranking

![image](https://user-images.githubusercontent.com/60184425/74457555-b2c51280-4e88-11ea-8446-b7ff3b60c027.png)

**Comment by koburk** [comment]
Full LOG

 
DB Object ranking_hist created
DB Object ranking created
DB Object error_check created
DB Object build_queue_process manages update_planet_values
DB Object update_planet_values (Stored procedure) created
DB Object reset_worl handles populate_world
DB Object populate_world (Stored procedure) created
DB Object reset_worl (Stored procedure) created
Random names for IA_Players -> Now using name_list instead
DB Object name_list created
DB Object build_queue_process (Stored procedure) created
DB Object update_planet_lastcheck (Stored procedure) created
DB Object set_new_object_levels (Stored procedure) created
DB Object ia_queue_to_general_queue (Stored procedure) created
DB Object ia_decide (Stored procedure) created
DB Object ia_build_hist created
DB Object set_new_object_levels (Stored procedure) created
DB Object update_planet_lastcheck (Stored procedure) created
DB Object build_queue created
DB Object ia_connection manages ia_brain object
DB Object ia_connection (Event) created
DB Object ia_brain manages queue_lock object
DB Object object_level created
DB Object wants_to_do (Function) created
DB Object ia_brain (Stored procedure) created
DB Object ia_queue created
DB Object is_online (Function) created
DB Object object_detail created
DB Object object_name created
DB Object universe created
DB Object rapid_fire created
DB Object requriments created
DB Object planets created
DB Object create_user manages full_galaxy object.
DB Object create_user (Stored procedure) created
DB Object change_log created
DB Object users created
DB Object ia_style_description created
DB Object types created
Random names for IA_Players
Change log created

**Comment by koburk** [comment]
I'm so proud, my lobotomized retarded IA Players are able to decide if they're gonna get energy from upgrading solar plants or buying satelites...
I just wanted to share it with you guys, I'm like a father for them ... haha

**Comment by gregariomansa** [comment]
Excellent! Good for you! Will you open your server for competition among
bots?

**Comment by koburk** [comment]
I will open the server for testing once my bots are less retarded and not all of them get lobotomized haha.
I need to find a way to give a bot (or group of bots) a goal, like "reaching the biggest metal mine" or stuff like that (like real people do).
Still i need further tests, i haven't developed any UI yet, so there's no way to "play" the game but being part of the ia. Like i said, i need longer days with more time to spend on this engine ...

Also, I've managed to give some sort of personality to each IA. Currently if you attack a player multiple times it will improve their defences and will "find a way" to not get attacked that much. This process was made because of 2 points.
1- The universe evolves and so should do them.
2- Farming something that has always the same thing is boring, we need to spy, simulate and then attack, not just take notes of a target list with coordenates (IMO).

**Comment by koburk** [comment]
:) Currently teaching my brain-dead lobotomized retarded IA_Players to colonize....
I'm splitting ia_players in agro or def style (basically), at the moment only defensive style is being used, i want them to colonize first and then i'll enable agro style to start teaching them to spy, then i need to build a simulation/battle engine, generate the math for each player (some will be a bit more reckless) to decide if attacking or not based on recources captured/recicled and ... yeah, i want the IA_PLAYERS to seek revenge!, so if you screw up with them they will find a way to get something in return... of course this could be something good to set baits and recicle their attacks ahaha (i need to give a small chance for that to happen but it will be feasible), at the end of the day the IA is there to push players economy so it won't be perfect...

Oh sh@t, i might need to recreate the fleet table structure to allow more stuff, like:
MOVE A PLANET. Oh yeah, what about using death stars to pull a planet from one place to another (it will consume huge amounts of deuterium) and will enable another mission to destroy those death stars pulling planets.
IMO this feature shouldn't be only for players with death stars but everyone, so you can send a fleet to hold position in a planet and then the owner of that planet can use those ships to move the planet for example, allowing other people to intercept the movement and recicle those poor and exhausted death stars ...
Still I need to think about some more missions, including renting, giving away...

**Comment by koburk** [comment]
Update

 - DB Object research_level removed -> Migrated to object_level instead
 - New research: Dark matter understanding. Each level will grant 1 unit of dark matter for every hour.
 - New ship: Driller. This ship will be able to drill within asteroids to gather resources.
 - New research: Drilling. It will allow material extraction from asteroids. LVL1 Metal, LVL4 Crystal, LVL7 Deuterium, LVL10 Dark matter.
 - New planet: Asteroid. It will allow drilling operations.
 - DB Object users_stk created
 - DB Object requirement_stk created
 - DB Object rapid_fire_stk created
 - DB Object planets_stk created
 - DB Object object_level_stk created
 - DB Object object_detail_stk created
 - DB Object full_galaxy_stk created
 - DB Object build_queue_stk created
 - DB Object create_events (Stored procedure) created
 - DB Object create_ranks (Stored procedure) created
 - DB Object move_to_disk (Stored procedure) created
 - DB Object randomize_name (Stored procedure) created
 - DB Object recover_to_memory (Stored procedure) created
 - DB Object task_daily (Stored procedure) created
 - DB Object try_to_buy_fleet (Stored procedure) created
 - DB Object try_to_fortify (Stored procedure) created
 - DB Object try_to_research (Stored procedure) created
 - DB Object is_building_def (Function) created
 - DB Object is_building_fleet (Function) created
 - DB Object is_researching (Function) created
 - DB Object try_to_build (Stored procedure) created
 - DB Object research_level created

**Comment by koburk** [comment]
Best design ever! haha, nah, but seriously, this is moving forward. Next steps will be enable sections to build structures -> researchs -> defences -> ships in that order.
After that i need to configure fleet orders and action for each order (Nightmare in progress!).

![image](https://user-images.githubusercontent.com/60184425/75690794-bdaede00-5ca3-11ea-8877-7ff8c79c5317.png)

**Comment by koburk** [comment]
Haha, i'm starting to get some stuff done. Im using ajax to avoid page reloads, feels smooth!

![image](https://user-images.githubusercontent.com/60184425/76089997-0d9ae700-5fbb-11ea-8d33-260dbc04b32a.png)

**Comment by koburk** [comment]
Researching is enabled

![image](https://user-images.githubusercontent.com/60184425/76229873-2c031b80-6223-11ea-8ad5-33afc73c3975.png)

**Comment by kroryan** [comment]
Si o guess u dont still working in this anymore right :(?
